### PR TITLE
parser: syntactically allow `self` in all `fn` contexts

### PR DESCRIPTION
**Pull request by Centril**
Part of https://github.com/rust-lang/rust/pull/68728.

`self` parameters are now *syntactically* allowed as the first parameter irrespective of item context (and in function pointers). Instead, semantic validation (`ast_validation`) is used.

r? @petrochenkov

### CODE REVIEW

File: src/librustc_parse/parser/item.rs [hunk 1]
```diff
@@ -1732,8 +1730,8 @@ impl<'a> Parser<'a> {
         attrs: Vec<Attribute>,
         header: FnHeader,
     ) -> PResult<'a, Option<P<Item>>> {
-        let (ident, decl, generics) =
-            self.parse_fn_sig(ParamCfg { is_self_allowed: false, is_name_required: |_| true })?;
+        let cfg = ParamCfg { is_name_required: |_| true };
```
Comment (Centril) [inline src/librustc_parse/parser/item.rs]: (I'm avoiding making changes to this now to make rebasing easier later.)

File: src/librustc_ast_passes/ast_validation.rs [hunk 2]
```diff
@@ -352,6 +366,21 @@ impl<'a> AstValidator<'a> {
...
+    fn check_decl_self_param(&self, fn_decl: &FnDecl, self_semantic: SelfSemantic) {
+        if let (SelfSemantic::No, [param, ..]) = (self_semantic, &*fn_decl.inputs) {
+            if param.is_self() {
+                self.err_handler()
+                    .struct_span_err(
+                        param.span,
+                        "`self` parameter is only allowed in associated functions",
+                    )
+                    .span_label(param.span, "not semantically valid as function parameter")
+                    .note("associated `fn`s are those in `impl` or `trait` definitions")
```
Comment (petrochenkov) [inline src/librustc_ast_passes/ast_validation.rs]:
```suggestion
                    .note("associated functions are those in `impl` or `trait` definitions")
```

**Comment by petrochenkov** [comment]
r=me with the wording tweak.

**Pull request merged**